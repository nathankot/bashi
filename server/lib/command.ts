export * from "./command/types.ts";
export * from "./command/parser.ts";
export * from "./command/builtinCommands.ts";
export * from "./command/filter.ts";
export * from "./command/run.ts";
export * from "./command/supportedModels.ts";
