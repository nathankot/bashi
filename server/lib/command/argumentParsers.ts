import * as t from "io-ts";
import { parseDate } from "chrono";

import { StringValue } from "@lib/valueTypes.ts";

export type ArgumentParserContext = {
  now: Date;
  chronoParseDate: typeof parseDate;
  timezoneUtcOffset: number;
};

export const argumentParsers = {
  naturalLanguageDateTime: {
    inputType: "string" as const,
    outputType: "string" as const,
    fn: function (ctx: ArgumentParserContext, arg: string): null | StringValue {
      const d = ctx.chronoParseDate(arg, {
        instant: ctx.now,
        timezone: ctx.timezoneUtcOffset,
      });
      if (d == null) {
        return null;
      }
      return {
        type: "string",
        value: d.toISOString(),
      };
    },
  },
};

export const ArgumentParser = t.keyof(argumentParsers);
export type ArgumentParser = t.TypeOf<typeof ArgumentParser>;

export default argumentParsers;
