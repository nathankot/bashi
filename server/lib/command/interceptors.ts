import math from "./interceptors/math.ts";
import time from "./interceptors/time.ts";
import translate from "./interceptors/translate.ts";
import generateCode from "./interceptors/generateCode.ts";
import editProse from "./interceptors/editProse.ts";
import editCode from "./interceptors/editCode.ts";

export const commandInterceptors = {
  math,
  time,
  translate,
  generateCode,
  editProse,
  editCode,
};
