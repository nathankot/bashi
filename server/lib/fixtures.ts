import type { Session } from "@lib/session.ts";
import type { CommandSet } from "@lib/command/types.ts";
import type { HandlerContext } from "$fresh/server.ts";
import builtinCommands from "@lib/command/builtinCommands.ts";

export const now = new Date(1671439270000);

export const handlerCtx: HandlerContext<{}> = {
  params: {},
  render: (data) => new Response(JSON.stringify(data)),
  renderNotFound: () => new Response("", { status: 404 }),
  state: {},
  localAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
  remoteAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
};

export const commandSet: CommandSet = {
  createCalendarEvent: {
    description: "create a calendar event on a certain date and time",
    args: [
      {
        name: "relative time supported by chrono-js",
        type: "string",
        parse: ["naturalLanguageDateTime"],
      },
      {
        name: "name",
        type: "string",
      },
    ],
    triggerTokens: ["calendar", "event"],
  },
  createReminder: {
    description: "create a reminder on a certain date and time",
    args: [
      {
        name: "relative time supported by chrono-js",
        type: "string",
        parse: ["naturalLanguageDateTime"],
      },
      { name: "name", type: "string" },
    ],
    triggerTokens: ["remind", "reminder"],
  },
  sendEmail: {
    description: `send an email`,
    args: [
      { name: "recipient", type: "string" },
      { name: "subject", type: "string" },
      { name: "contents", type: "string" },
    ],
    triggerTokens: ["email"],
  },
};

export const session: Session = {
  expiresAt: new Date(new Date().getTime() + 60000),
  sessionId: "a9fb6273-00ee-4e4c-9918-e87e1157ca31",
  accountNumber: "0000000000000000",
  configuration: {
    locale: "en-US",
    maxResponseTokens: 1000,
    bestOf: 1,
    enabledBuiltinCommands: Object.keys(builtinCommands) as any,
    timezoneUtcOffset: 0,
  },
  modelConfigurations: {
    "assist-000": {
      model: "assist-000",
      commands: commandSet,
    },
  },
};
