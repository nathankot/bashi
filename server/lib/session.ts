import * as t from "io-ts";
import { date } from "io-ts-types";

import { AllConfiguration as ModelConfiguration } from "@lib/models.ts";
import {
  Configuration,
  defaultConfiguration,
} from "@lib/session/configuration.ts";

export const Session = t.type({
  sessionId: t.string,
  expiresAt: date,
  configuration: Configuration,
  modelConfigurations: t.array(ModelConfiguration),
});

export type Session = t.TypeOf<typeof Session>;

export { defaultConfiguration };
