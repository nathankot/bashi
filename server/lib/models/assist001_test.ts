import { noop } from "cockatiel";

import { assertSnapshot } from "std/testing/snapshot.ts";
import * as fixtures from "@lib/fixtures.ts";
import { Session } from "@lib/session.ts";
import { CommandExecuted, parseFunctionCall } from "@lib/command.ts";

import {
  MAX_MODEL_CALLS,
  Input,
  run,
  getPendingCommandsOrResult,
} from "./assist001.ts";

const pendingClientCommandState = {
  modelCallCount: 1,
  pending: {
    action: 'now(); ask("what do you want?")',
    functionCalls: [
      { type: "call", args: [], name: "now" },
      {
        type: "call",
        args: [{ type: "string", value: "what do you want?" }],
        name: "ask",
      },
    ],
    result: undefined,
    thought: "I need to do something",
  },
  request: "some request",
  requestContext: {},
  resolvedActionGroups: [],
  resolvedCommands: {
    "0.0": {
      args: [],
      id: "0",
      name: "now",
      returnValue: { type: "string", value: "2022-12-19T08:41:10.000Z" },
      type: "executed",
    },
  },
} satisfies Session["assist001State"];

const pendingRequestContextState = {
  modelCallCount: 1,
  pending: {
    action: 'editProse("convert to poem"); now()',
    functionCalls: [
      {
        type: "call",
        args: [{ type: "string", value: "convert to poem" }],
        name: "editProse",
      },
      {
        type: "call",
        args: [],
        name: "now",
      },
    ],
    result: undefined,
    thought: "I need to do something",
  },
  request: "some request",
  requestContext: {},
  resolvedActionGroups: [
    {
      action: "someCommand()",
      result: `"blah"`,
      thought: "I need to call some command",
      functionCalls: [
        {
          type: "call",
          args: [],
          name: "someCommand",
        },
      ],
    },
  ],
  resolvedCommands: {
    "0.0": {
      type: "executed",
      args: [],
      id: "0",
      name: "someCommand",
      returnValue: { type: "string", value: "blah" },
    },
  },
} satisfies Session["assist001State"];

for (const test of [
  {
    description: "all commands resolved on the server",
    input: { request: "some request" },
    openAiResults: [
      `I need to do something
Action: now(); math("pi^2 + 123")`,
      `I have finished
Action: finish()`,
    ],
    snapshotPrompts: true,
  },
  {
    description: "all commands resolved on the server - implicit finish",
    input: { request: "some request" },
    openAiResults: [
      `I need to do something
Action: now(); math("pi^2 + 123")`,
      ``,
    ],
  },
  {
    description: "client resolved command",
    input: { request: "some request" },
    openAiResults: [
      `I need to do something
Action: now(); ask("what do you want?")`,
    ],
  },
  {
    description: "client resolved command - continue but unresolved",
    input: {
      resolvedCommands: {},
    },
    openAiResults: [],
    initialState: pendingClientCommandState,
  },
  {
    description: "client resolved command - wrong return type",
    input: {
      resolvedCommands: { 1: { type: "number", value: 123 } },
    },
    openAiResults: [],
    snapshotError: true,
    initialState: pendingClientCommandState,
  },
  {
    description: "client resolved command - fulfilled",
    input: {
      resolvedCommands: { 1: { type: "string", value: "to test you" } },
    },
    openAiResults: [`Finished\nAction: finish()`],
    snapshotPrompts: true,
    initialState: pendingClientCommandState,
  },
  {
    description: "nested calls",
    input: { request: "some request" },
    openAiResults: [
      `I need to get the current time in New York and create a calendar event 5 days from now
Action: time("America/New_York"); createCalendarEvent(relativeTime("5 days from now"), "Dinner with Wife")`,
    ],
    snapshotPrompts: true,
  },
  {
    description: "request needs more context",
    input: { request: "some request" },
    openAiResults: [
      `I need to do something
Action: now(); editProse("convert to poem"); now()`,
    ],
  },
  {
    description: "request needs more context - still missing",
    input: { requestContext: {} },
    openAiResults: [],
    initialState: pendingRequestContextState,
  },
  {
    description: "request needs more context - wrong type",
    input: { requestContext: { text: { type: "number", value: 123 } } },
    openAiResults: [],
    snapshotError: true,
    initialState: pendingRequestContextState,
  },
  {
    description: "request needs more context - fulfilled",
    input: { requestContext: { text: { type: "string", value: `some text` } } },
    openAiResults: [
      `the result of editProse()`,
      `I am finished\nAction: finish()`,
    ],
    snapshotPrompts: true,
    initialState: pendingRequestContextState,
  },
  {
    description: "fulfilled but max loops",
    input: { requestContext: { text: { type: "string", value: `some text` } } },
    openAiResults: [`the result of editProse()`],
    snapshotError: true,
    initialState: {
      ...pendingRequestContextState,
      modelCallCount: MAX_MODEL_CALLS,
    },
  },
  {
    description: "wrong arg type",
    input: { request: "some request" },
    openAiResults: [
      `I need to do something
Action: now(); math(123)`,
    ],
    snapshotError: true,
  },
  {
    description: "wrong arg count",
    input: { request: "some request" },
    openAiResults: [
      `I need to do something
Action: now(); math()`,
    ],
    snapshotError: true,
  },

  // model uses wrong arg types
  // model uses wrong arg count
  // TODO: test arg parsers (natural language)
] as {
  description: string;
  openAiResults?: string[];
  input: Input;
  initialState?: NonNullable<Session["assist001State"]>;
  snapshotPrompts?: true;
  snapshotError?: true;
}[]) {
  Deno.test(test.description, async (t) => {
    let n = 0;
    let session: Session = {
      ...fixtures.session,
      assist001State: test.initialState,
    };
    let prompts: string[] = [];

    const openAiClient = {
      createCompletion(opts: { prompt: [string] }) {
        prompts.push(opts.prompt[0]);
        const text = (test.openAiResults ?? [])[n];
        if (text == null) {
          throw new Error(`openai mock on index ${n} not available`);
        }
        n++;
        return { data: { choices: [{ text }] } };
      },
    };

    try {
      const output = await run(
        {
          faultHandlingPolicy: noop,
          log: () => {},
          now: () => fixtures.now,
          openai: openAiClient as any,
          session: session,
          setUpdatedSession: (s) => {
            session = s;
          },
          signal: null as any,
          whisperEndpoint: null as any,
        },
        {
          model: "assist-001",
          commands: fixtures.commandSet,
        },
        test.input
      );

      await assertSnapshot(t, {
        output,
        state: session?.assist001State,
        n,
      });

      if (test.snapshotPrompts === true) {
        await assertSnapshot(t, prompts);
      }
    } catch (e) {
      if (test.snapshotError !== true) {
        throw e;
      }

      await assertSnapshot(t, e.message);
    }
  });
}

for (const test of [
  {
    description: "example A step 1",
    commandId: "0",
    call: parseFunctionCall(`test(a(), b(123, c()))`),
    resolvedCommands: {},
  },
  {
    description: "example A step 2",
    commandId: "0",
    call: parseFunctionCall(`test(a(), b(123, c()))`),
    resolvedCommands: {
      "0.0": {
        type: "executed",
        args: [],
        id: "0.0",
        name: "a",
        returnValue: { type: "number", value: 123 },
      },
      "0.1.1": {
        type: "executed",
        args: [],
        id: "0.1.1",
        name: "c",
        returnValue: { type: "string", value: "blah" },
      },
    } as Record<string, CommandExecuted>,
  },
  {
    description: "example A step 3",
    commandId: "0",
    call: parseFunctionCall(`test(a(), b(123, c()))`),
    resolvedCommands: {
      "0.0": {
        type: "executed",
        args: [],
        id: "0.0",
        name: "a",
        returnValue: { type: "number", value: 123 },
      },
      "0.1": {
        type: "executed",
        args: [],
        id: "0.1",
        name: "b",
        returnValue: { type: "string", value: "ha" },
      },
    } as Record<string, CommandExecuted>,
  },
  {
    description: "example A step 4",
    commandId: "0",
    call: parseFunctionCall(`test(a(), b(123, c()))`),
    resolvedCommands: {
      "0": {
        type: "executed",
        args: [],
        id: "0",
        name: "test",
        returnValue: { type: "number", value: 123 },
      },
    } as Record<string, CommandExecuted>,
  },
]) {
  Deno.test(test.description, async (t) => {
    try {
      await assertSnapshot(
        t,
        getPendingCommandsOrResult(
          test.commandId,
          test.call,
          test.resolvedCommands
        )
      );
    } catch (e) {
      await assertSnapshot(t, e.message);
    }
  });
}
