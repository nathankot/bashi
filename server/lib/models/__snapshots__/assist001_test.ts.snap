export const snapshot = {};

snapshot[`all commands resolved on the server 1`] = `
{
  n: 2,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      resolvedCommands: {
        "0.0": {
          args: [
          ],
          id: "0.0",
          name: "now",
          returnValue: {
            type: "string",
            value: "2022-12-19T08:41:10.000Z",
          },
          type: "executed",
        },
        "0.1": {
          args: [
            {
              type: "string",
              value: "pi^2 + 123",
            },
          ],
          id: "0.1",
          name: "math",
          returnValue: {
            type: "string",
            value: "132.86960440108936",
          },
          type: "executed",
        },
        "1.0": {
          args: [
          ],
          id: "1.0",
          name: "finish",
          returnValue: {
            type: "null",
          },
          type: "executed",
        },
      },
      type: "finished",
    },
  },
  state: undefined,
}
`;

snapshot[`all commands resolved on the server 2`] = `
[
  'Answer the following questions as directly and as best you can.

You have access to tools denoted below in Typescript-like function declarations. String arguments MUST be quoted and any quotes inside them MUST be escaped. Functions that are not listed below MUST NOT be used. Function arguments MUST be literal types. Function calls MAY be nested:

\`answer(answer: string) => null\` - answer the original question directly
\`say(message: string) => null\` - say something
\`write(message: string) => null\` - write/insert a message
\`ask(question: string) => string\` - ask for additional information
\`now() => string\` - get the current time in ISO8601 format
\`math("a mathjs expression ": string) => string\` - compute a math formula
\`relativeTime("natural language description of relative time": string) => string\` - get the time relative to now in ISO8601 format
\`finish() => null\` - mark that the request is finished
\`fail(reason: string) => null\` - indicate the request could not be interpreted

Use the following format:

Request: the input question or request you must answer
Thought: you should always think about what to do
Action: function(s) to call, delimited by ;
Result: the result of the function call
... (this Thought/Action/Result can repeat N times)
Thought: I have completed the request
Action: finish()

Begin!

Request: some request
Thought: ',
  'Answer the following questions as directly and as best you can.

You have access to tools denoted below in Typescript-like function declarations. String arguments MUST be quoted and any quotes inside them MUST be escaped. Functions that are not listed below MUST NOT be used. Function arguments MUST be literal types. Function calls MAY be nested:

\`answer(answer: string) => null\` - answer the original question directly
\`say(message: string) => null\` - say something
\`write(message: string) => null\` - write/insert a message
\`ask(question: string) => string\` - ask for additional information
\`now() => string\` - get the current time in ISO8601 format
\`math("a mathjs expression ": string) => string\` - compute a math formula
\`relativeTime("natural language description of relative time": string) => string\` - get the time relative to now in ISO8601 format
\`finish() => null\` - mark that the request is finished
\`fail(reason: string) => null\` - indicate the request could not be interpreted

Use the following format:

Request: the input question or request you must answer
Thought: you should always think about what to do
Action: function(s) to call, delimited by ;
Result: the result of the function call
... (this Thought/Action/Result can repeat N times)
Thought: I have completed the request
Action: finish()

Begin!

Request: some request
Thought: I need to do something
Action: now(); math("pi^2 + 123")
Result: "2022-12-19T08:41:10.000Z"; "132.86960440108936"
Thought: ',
]
`;

snapshot[`all commands resolved on the server - implicit finish 1`] = `
{
  n: 2,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      resolvedCommands: {
        "0.0": {
          args: [
          ],
          id: "0.0",
          name: "now",
          returnValue: {
            type: "string",
            value: "2022-12-19T08:41:10.000Z",
          },
          type: "executed",
        },
        "0.1": {
          args: [
            {
              type: "string",
              value: "pi^2 + 123",
            },
          ],
          id: "0.1",
          name: "math",
          returnValue: {
            type: "string",
            value: "132.86960440108936",
          },
          type: "executed",
        },
      },
      type: "finished",
    },
  },
  state: undefined,
}
`;

snapshot[`client resolved command 1`] = `
{
  n: 1,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      pendingCommands: [
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          id: "0.1",
          name: "ask",
          type: "parsed",
        },
      ],
      type: "pending_commands",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'now(); ask("what do you want?")',
      functionCalls: [
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          name: "ask",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to do something",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
    ],
    resolvedCommands: {
      "0.0": {
        args: [
        ],
        id: "0.0",
        name: "now",
        returnValue: {
          type: "string",
          value: "2022-12-19T08:41:10.000Z",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`client resolved command - continue but unresolved 1`] = `
{
  n: 0,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      pendingCommands: [
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          id: "0.1",
          name: "ask",
          type: "parsed",
        },
      ],
      type: "pending_commands",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'now(); ask("what do you want?")',
      functionCalls: [
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          name: "ask",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to do something",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
    ],
    resolvedCommands: {
      "0.0": {
        args: [
        ],
        id: "0",
        name: "now",
        returnValue: {
          type: "string",
          value: "2022-12-19T08:41:10.000Z",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`client resolved command - wrong return type 1`] = `
{
  n: 0,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      pendingCommands: [
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          id: "0.1",
          name: "ask",
          type: "parsed",
        },
      ],
      type: "pending_commands",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'now(); ask("what do you want?")',
      functionCalls: [
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          name: "ask",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to do something",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
    ],
    resolvedCommands: {
      "0.0": {
        args: [
        ],
        id: "0",
        name: "now",
        returnValue: {
          type: "string",
          value: "2022-12-19T08:41:10.000Z",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`client resolved command - fulfilled 1`] = `
{
  n: 0,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      pendingCommands: [
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          id: "0.1",
          name: "ask",
          type: "parsed",
        },
      ],
      type: "pending_commands",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'now(); ask("what do you want?")',
      functionCalls: [
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
        {
          args: [
            {
              type: "string",
              value: "what do you want?",
            },
          ],
          name: "ask",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to do something",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
    ],
    resolvedCommands: {
      "0.0": {
        args: [
        ],
        id: "0",
        name: "now",
        returnValue: {
          type: "string",
          value: "2022-12-19T08:41:10.000Z",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`client resolved command - fulfilled 2`] = `
[
]
`;

snapshot[`nested calls 1`] = `
{
  n: 1,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      pendingCommands: [
        {
          args: [
            {
              type: "string",
              value: "2022-12-24T08:41:10.000Z",
            },
            {
              type: "string",
              value: "Dinner with Wife",
            },
          ],
          id: "0.1",
          name: "createCalendarEvent",
          type: "parsed",
        },
      ],
      type: "pending_commands",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'time("America/New_York"); createCalendarEvent(relativeTime("5 days from now"), "Dinner with Wife")',
      functionCalls: [
        {
          args: [
            {
              type: "string",
              value: "America/New_York",
            },
          ],
          name: "time",
          type: "call",
        },
        {
          args: [
            {
              args: [
                {
                  type: "string",
                  value: "5 days from now",
                },
              ],
              name: "relativeTime",
              type: "call",
            },
            {
              type: "string",
              value: "Dinner with Wife",
            },
          ],
          name: "createCalendarEvent",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to get the current time in New York and create a calendar event 5 days from now",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
    ],
    resolvedCommands: {
      "0.0": {
        args: [
          {
            type: "string",
            value: "America/New_York",
          },
        ],
        id: "0.0",
        name: "time",
        returnValue: {
          type: "string",
          value: "12/19/2022, 3:41:10 AM",
        },
        type: "executed",
      },
      "0.1.0": {
        args: [
          {
            type: "string",
            value: "5 days from now",
          },
        ],
        id: "0.1.0",
        name: "relativeTime",
        returnValue: {
          type: "string",
          value: "2022-12-24T08:41:10.000Z",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`nested calls 2`] = `
[
  'Answer the following questions as directly and as best you can.

You have access to tools denoted below in Typescript-like function declarations. String arguments MUST be quoted and any quotes inside them MUST be escaped. Functions that are not listed below MUST NOT be used. Function arguments MUST be literal types. Function calls MAY be nested:

\`answer(answer: string) => null\` - answer the original question directly
\`say(message: string) => null\` - say something
\`write(message: string) => null\` - write/insert a message
\`ask(question: string) => string\` - ask for additional information
\`now() => string\` - get the current time in ISO8601 format
\`math("a mathjs expression ": string) => string\` - compute a math formula
\`relativeTime("natural language description of relative time": string) => string\` - get the time relative to now in ISO8601 format
\`finish() => null\` - mark that the request is finished
\`fail(reason: string) => null\` - indicate the request could not be interpreted

Use the following format:

Request: the input question or request you must answer
Thought: you should always think about what to do
Action: function(s) to call, delimited by ;
Result: the result of the function call
... (this Thought/Action/Result can repeat N times)
Thought: I have completed the request
Action: finish()

Begin!

Request: some request
Thought: ',
]
`;

snapshot[`request needs more context 1`] = `
{
  n: 1,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      missingRequestContext: {
        text: {
          type: "string",
        },
      },
      type: "needs_request_context",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'now(); editProse("convert to poem"); now()',
      functionCalls: [
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
        {
          args: [
            {
              type: "string",
              value: "convert to poem",
            },
          ],
          name: "editProse",
          type: "call",
        },
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to do something",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
    ],
    resolvedCommands: {
      "0.0": {
        args: [
        ],
        id: "0.0",
        name: "now",
        returnValue: {
          type: "string",
          value: "2022-12-19T08:41:10.000Z",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`request needs more context - still missing 1`] = `
{
  n: 0,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      missingRequestContext: {
        text: {
          type: "string",
        },
      },
      type: "needs_request_context",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'editProse("convert to poem"); now()',
      functionCalls: [
        {
          args: [
            {
              type: "string",
              value: "convert to poem",
            },
          ],
          name: "editProse",
          type: "call",
        },
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to do something",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
      {
        action: "someCommand()",
        functionCalls: [
          {
            args: [
            ],
            name: "someCommand",
            type: "call",
          },
        ],
        result: '"blah"',
        thought: "I need to call some command",
      },
    ],
    resolvedCommands: {
      "0.0": {
        args: [
        ],
        id: "0",
        name: "someCommand",
        returnValue: {
          type: "string",
          value: "blah",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`request needs more context - wrong type 1`] = `
{
  n: 0,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      missingRequestContext: {
        text: {
          type: "string",
        },
      },
      type: "needs_request_context",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'editProse("convert to poem"); now()',
      functionCalls: [
        {
          args: [
            {
              type: "string",
              value: "convert to poem",
            },
          ],
          name: "editProse",
          type: "call",
        },
        {
          args: [
          ],
          name: "now",
          type: "call",
        },
      ],
      result: undefined,
      thought: "I need to do something",
    },
    request: "some request",
    requestContext: {
      text: {
        type: "number",
        value: 123,
      },
    },
    resolvedActionGroups: [
      {
        action: "someCommand()",
        functionCalls: [
          {
            args: [
            ],
            name: "someCommand",
            type: "call",
          },
        ],
        result: '"blah"',
        thought: "I need to call some command",
      },
    ],
    resolvedCommands: {
      "0.0": {
        args: [
        ],
        id: "0",
        name: "someCommand",
        returnValue: {
          type: "string",
          value: "blah",
        },
        type: "executed",
      },
    },
  },
}
`;

snapshot[`request needs more context - fulfilled 1`] = `
{
  n: 2,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      resolvedCommands: {
        "0.0": {
          args: [
          ],
          id: "0",
          name: "someCommand",
          returnValue: {
            type: "string",
            value: "blah",
          },
          type: "executed",
        },
        "1.0": {
          args: [
            {
              type: "string",
              value: "convert to poem",
            },
          ],
          id: "1.0",
          name: "editProse",
          returnValue: {
            type: "string",
            value: "the result of editProse()",
          },
          type: "executed",
        },
        "1.1": {
          args: [
          ],
          id: "1.1",
          name: "now",
          returnValue: {
            type: "string",
            value: "2022-12-19T08:41:10.000Z",
          },
          type: "executed",
        },
        "2.0": {
          args: [
          ],
          id: "2.0",
          name: "finish",
          returnValue: {
            type: "null",
          },
          type: "executed",
        },
      },
      type: "finished",
    },
  },
  state: undefined,
}
`;

snapshot[`request needs more context - fulfilled 2`] = `
[
  "Rewrite and edit the following text. The requirement is 'convert to poem':

some text",
  'Answer the following questions as directly and as best you can.

You have access to tools denoted below in Typescript-like function declarations. String arguments MUST be quoted and any quotes inside them MUST be escaped. Functions that are not listed below MUST NOT be used. Function arguments MUST be literal types. Function calls MAY be nested:

\`answer(answer: string) => null\` - answer the original question directly
\`say(message: string) => null\` - say something
\`write(message: string) => null\` - write/insert a message
\`ask(question: string) => string\` - ask for additional information
\`now() => string\` - get the current time in ISO8601 format
\`math("a mathjs expression ": string) => string\` - compute a math formula
\`relativeTime("natural language description of relative time": string) => string\` - get the time relative to now in ISO8601 format
\`finish() => null\` - mark that the request is finished
\`fail(reason: string) => null\` - indicate the request could not be interpreted

Use the following format:

Request: the input question or request you must answer
Thought: you should always think about what to do
Action: function(s) to call, delimited by ;
Result: the result of the function call
... (this Thought/Action/Result can repeat N times)
Thought: I have completed the request
Action: finish()

Begin!

Request: some request
Thought: I need to call some command
Action: someCommand()
Result: "blah"
Thought: I need to do something
Action: editProse("convert to poem"); now()
Result: "the result of editProse()"; "2022-12-19T08:41:10.000Z"
Thought: ',
]
`;

snapshot[`fulfilled but max loops 1`] = `"max iteration count of 5 reached"`;

snapshot[`wrong arg type 1`] = `"arguments are invalid"`;

snapshot[`wrong arg count 1`] = `"arguments are invalid"`;

snapshot[`top level commands are resolved sequentially 1`] = `
{
  n: 1,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      pendingCommands: [
        {
          args: [
            {
              type: "string",
              value: "how are you?",
            },
          ],
          id: "0.0",
          name: "ask",
          type: "parsed",
        },
      ],
      type: "pending_commands",
    },
  },
  state: {
    modelCallCount: 1,
    pending: {
      action: 'ask("how are you?"); time("America/New_York")',
      functionCalls: [
        {
          args: [
            {
              type: "string",
              value: "how are you?",
            },
          ],
          name: "ask",
          type: "call",
        },
        {
          args: [
            {
              type: "string",
              value: "America/New_York",
            },
          ],
          name: "time",
          type: "call",
        },
      ],
      result: undefined,
      thought: "some thought",
    },
    request: "some request",
    requestContext: {},
    resolvedActionGroups: [
    ],
    resolvedCommands: {},
  },
}
`;

snapshot[`top level commands are resolved sequentially 2 1`] = `
{
  n: 1,
  output: {
    model: "assist-001",
    request: "some request",
    result: {
      resolvedCommands: {
        "0.0": {
          args: [
            {
              type: "string",
              value: "how are you?",
            },
          ],
          id: "0.0",
          name: "ask",
          returnValue: {
            type: "string",
            value: "good",
          },
          type: "executed",
        },
        "0.1": {
          args: [
            {
              type: "string",
              value: "America/New_York",
            },
          ],
          id: "0.1",
          name: "time",
          returnValue: {
            type: "string",
            value: "12/19/2022, 3:41:10 AM",
          },
          type: "executed",
        },
        "1.0": {
          args: [
          ],
          id: "1.0",
          name: "finish",
          returnValue: {
            type: "null",
          },
          type: "executed",
        },
      },
      type: "finished",
    },
  },
  state: undefined,
}
`;

snapshot[`example A step 1 1`] = `
{
  pendingCommands: [
    {
      args: [
      ],
      id: "0.0",
      name: "a",
      type: "parsed",
    },
    {
      args: [
      ],
      id: "0.1.1",
      name: "c",
      type: "parsed",
    },
  ],
}
`;

snapshot[`example A step 2 1`] = `
{
  pendingCommands: [
    {
      args: [
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "blah",
        },
      ],
      id: "0.1",
      name: "b",
      type: "parsed",
    },
  ],
}
`;

snapshot[`example A step 3 1`] = `
{
  pendingCommands: [
    {
      args: [
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "ha",
        },
      ],
      id: "0",
      name: "test",
      type: "parsed",
    },
  ],
}
`;

snapshot[`example A step 4 1`] = `
{
  result: {
    type: "number",
    value: 123,
  },
}
`;
